### ISSUE TITLE
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa2: invalid start byte for some org files

### ISSUE THREAD

**Issue by emacsomancer** [debanjum/khoj#89]
khoj fails to index a directory of org files, producing error:

```
Traceback (most recent call last):
  File "/home/slade/.local/bin/khoj", line 8, in <module>
    sys.exit(run())
  File "/home/slade/.local/lib/python3.10/site-packages/src/main.py", line 112, in run
    configure_server(args, required=False)
  File "/home/slade/.local/lib/python3.10/site-packages/src/configure.py", line 36, in configure_server
    state.model = configure_search(state.model, state.config, args.regenerate)
  File "/home/slade/.local/lib/python3.10/site-packages/src/configure.py", line 46, in configure_search
    model.orgmode_search = text_search.setup(
  File "/home/slade/.local/lib/python3.10/site-packages/src/search_type/text_search.py", line 173, in setup
    text_to_jsonl(config.input_files, config.input_filter, config.compressed_jsonl)
  File "/home/slade/.local/lib/python3.10/site-packages/src/processor/org_mode/org_to_jsonl.py", line 32, in org_to_jsonl
    entries, file_to_entries = extract_org_entries(org_files)
  File "/home/slade/.local/lib/python3.10/site-packages/src/processor/org_mode/org_to_jsonl.py", line 72, in extract_org_entries
    org_file_entries = orgnode.makelist(str(org_file))
  File "/home/slade/.local/lib/python3.10/site-packages/src/processor/org_mode/orgnode.py", line 83, in makelist
    for line in f:
  File "/usr/lib/python3.10/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa2 in position 676: invalid start byte
```

I'm not sure which files are triggering this error, but will experiment.

**Comment by debanjum** [comment]
Hey @emacsomancer, some pointers below to find, fix the issue you're facing:
1. Run `grep -axv '.*' /path/to/notes/*.org` to find files with invalid utf-8 characters
   Refer: [StackOverflow](https://stackoverflow.com/a/41741313/19720296) answer for details on above command
2. Try install, run the development version of the app generated for #91. Check if that resolves your issue
   ```shell
   pip install -i https://test.pypi.org/simple/ khoj-assistant==0.1.9.dev911662905500 && khoj
   ```

**Comment by emacsomancer** [comment]
I was able to locate the files that had invalid utf-8 characters in them. but the development version of khoj still produced the same errors as long as the characters were in the files.

**Comment by debanjum** [comment]
I see. Have you been able to get `khoj` to work for you yet? If you can share the portion of the file with the invalid utf-8 characters it'll help resolve the issue faster

**Comment by debanjum** [comment]
Closing this as the issue has become stale. Suggested fix for the decoding error could not be validated. Feel free to reopen if anyone faces the same issue

**Comment by tleers** [comment]
I ran into this same issue, changing the following line in processor/markdown/markdown_to_jsonl.py fixed the issue for me.

```python
for markdown_file in markdown_files:
    with open(markdown_file) as f:
        markdown_content = f.read()
```
to 
```python
for markdown_file in markdown_files:
    with open(markdown_file, encoding='utf8', errors='ignore') as f:
        markdown_content = f.read()
```
But I'm not sure how desirable ignoring such errors would be.

**Comment by debanjum** [comment]
Hey @tleers, thanks a ton for looking into a potential code fix for this issue! I'd suggested the same [fix](https://github.com/debanjum/khoj/pull/91) for reading org-mode files. But that hadn't resolved the issue it seemed.

> But I'm not sure how desirable ignoring such errors would be.

Yeah, I'm not sure if that's a good idea either

It'd be great if you can provide some details to help fix this issue:
1. What is the exact `UnicodeDecodeError` you are/were seeing on your end? 
   [Abhuva79](https://www.reddit.com/r/ObsidianMD/comments/10thrpl/comment/j77iwda/) saw a `UnicodeDecodeError: 'charmap' codec can't decode byte...`, but [emacsomancer](https://github.com/debanjum/khoj/issues/89#issue-1368766506) was seeing `UnicodeDecodeError: 'utf-8' codec can't decode byte...`. These maybe 2 different issues
2. Do you know what your markdown files are encoded as? Are they all encoded in the same format?
3. Are you able to get any results from the files which are encoded in non utf8 format with the above fix?
4. Does opening the file as `utf8` without ignoring the errors still fix the issue for you? If so, I'd rather go with that as a fix for now
    ```python
    for markdown_file in markdown_files:
        with open(markdown_file, encoding='utf8') as f:
            markdown_content = f.read()
    ```

**Comment by sabaimran** [comment]
Steps to reproduce the original error:
```
touch test.md
vim test.md
# Add some Markdown content to the file and save
iconv -t UTF-16 test.md > test-encoding.md
python3 -m src.main -vv
```

Observed result: `test.md` has embeddings generated appropriately, while `test-encoding.md` errors out with `UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte`.

The root cause of the error is that the plain text files are encoded in a non-utf8 format. There's some reading [here](http://farmdev.com/talks/unicode/) and [here](https://www.joelonsoftware.com/2003/10/08/the-absolute-minimum-every-software-developer-absolutely-positively-must-know-about-unicode-and-character-sets-no-excuses/) about different encoding formats, what they mean, and how they're used in Python. 

Ignoring the errors is probably not a good solution because it causes garbled text output for characters that aren't recognized, though it does unblock users from using `khoj`. In that case, many results would look something like "?????". You can see guidance [here](https://docs.python.org/3/library/functions.html#open) in the Python API for different options for a decoding error in the `open` method.

One potential solution here is to use the `strict` errors resolution in the `open` method and apply `try-catches` with the various encoding formats that are most common. For `codecs` to be able to decode the file in the first place, it has to know the source encoding. With the source encoding, you should be able to [convert to utf-8](https://stackoverflow.com/questions/191359/how-to-convert-a-file-to-utf-8-in-python).

It probably warrants further investigation to see what other solutions exist to this problem, because I imagine such issues must be common.

**Comment by debanjum** [comment]
Marking this issue as resolved. This issue was effectively tracking 3 separate issues:
1. File reads were using default system encoding to open files. But obsidian encodes all markdown files as utf-8. This was causing the `UnicodeDecodeError`. Fixed this by reading markdown and jsonl as utf-8 in https://github.com/debanjum/khoj/commit/99a03da3f79eed24e5a7dcef3d2b15faa0926366
3. The previous solution triggered another bug as the jsonl index writer wasn't (always) writing to the index as utf-8. This was causing the read of the jsonl as utf-8 to fail. This was fixed in https://github.com/debanjum/khoj/commit/11517ba8eb3a968f1100caaedd3043d9310cf230
4. The more general issue of a mismatch in the file encoding and the system default encoding causing the decode error. To solve this, we'd need a more complex solution. Potential options are:
    1. Relying on the [chardet](https://chardet.readthedocs.io/en/latest/usage.html) to infer the potential encoding before read
    2. The try/catch method mentioned in [a previous comment](https://github.com/debanjum/khoj/issues/89#issuecomment-1419300030)
  
   If enough users complain about this, I'll open a new issue for this. 
   But for now the previous fixes should work for all Obsidian users of Khoj at least

*Note: Fixes `1.` and `2.` has gone through some validation based on discussions with users facing this issue*
